File: src/lib/notify.js
import { exec as execCb } from "node:child_process";
import { promisify } from "node:util";

const exec = promisify(execCb);

async function installDeps() {
  const { stderr } = await exec("brew install terminal-notifier");

  if (stderr) {
    throw new Error(stderr);
  }
}

async function hasNecessaryDeps() {
  const { stdout, stderr } = await exec("brew list | grep terminal-notifier");

  if (stderr) {
    throw new Error(stderr);
  }

  return !!stdout;
}

async function prepare() {
  const hasDeps = await hasNecessaryDeps();

  if (!hasDeps) {
    await installDeps();
  }
}

async function notify({ title, icon, message, link }) {
  try {
    const command = `terminal-notifier -message "${message}" -title "${title}" -open "${link}" -sound default`;

    await prepare();
    await exec(command);
  } catch {}
}

export default notify;

File: src/lib/notify.py
import os
from entities.ad import Ad


def __install_deps() -> None:
    os.system("brew install terminal-notifier")


def __has_necessary_deps() -> bool:
   # TODO get stdout
    stdout = ""
    os.system("brew list | grep terminal-notifier")

    return bool(stdout)


def __notify(
    icon: str,
    link: str,
    message: str,
    title: str
):
    app_icon = "src/imgs/olx.png"
    command = f"terminal-notifier -appIcon {app_icon} -message \"{message}\" -title \"{
        title}\" -open \"{link}\" -contentImage \"{icon}\" -sound default"

    if not __has_necessary_deps():
        __install_deps()

    os.system(command)


def notify(term: str, ads: list[Ad]):
    for ad in ads:
        __notify(
            icon=ad.thumbnail,
            link=ad.url,
            message=f"{ad.price} - {ad.title}",
            title=f"New offer for {term}"
        )
